### PR TITLE
Introduce Concurrent Merkle Tree Implementation

### PR DESCRIPTION
**Pull request by ngundotra**
Add rust libraries to enable concurrent modification of merkle trees by fast-forwarding merkle proofs that were valid for older states of the tree. We refer to this data structure as a "**Concurrent Merkle Tree.**"

The Concurrent Merkle Tree implementation is currently used to support compression of on-chain data. This program source code will be included in a separate PR.

Information about the Concurrent Merkle Tree data structure can be found in the [whitepaper](https://drive.google.com/file/d/1BOpa5OFmara50fTvL0VIVYjtg-qzHCVc/view). This includes a table comparing cost of native on-chain storage to the cost of verifying data in an on-chain Concurrent Merkle Tree.

This code is already in-use in a contract called Gummyroll, deployed on [mainnet-beta](https://explorer.solana.com/address/GRoLLzvxpxxu2PGNJMMeZPyMxjAUH9pKqxGXV9DGiceU), [devnet](https://explorer.solana.com/address/GRoLLzvxpxxu2PGNJMMeZPyMxjAUH9pKqxGXV9DGiceU?cluster=devnet), and [testnet](https://explorer.solana.com/address/GRoLLzvxpxxu2PGNJMMeZPyMxjAUH9pKqxGXV9DGiceU?cluster=testnet).

### CODE REVIEW

File: libraries/concurrent-merkle-tree/Cargo.toml
Comment (jarry-xiao) [inline libraries/concurrent-merkle-tree/Cargo.toml:19]: This should change to a crate once we get everything merged in

File: libraries/concurrent-merkle-tree/Cargo.toml
Comment (jarry-xiao) [inline libraries/concurrent-merkle-tree/Cargo.toml:13]: I don't think sol-log should be the default

File: libraries/concurrent-merkle-tree/src/state.rs
Comment (joncinque) [inline libraries/concurrent-merkle-tree/src/state.rs:80]: Since you've defined `Node` as 32 bytes, can you safely derive Pod / Zeroable on this?

File: libraries/concurrent-merkle-tree/src/cmt.rs
Comment (joncinque) [inline libraries/concurrent-merkle-tree/src/cmt.rs:42]: I'm not sure if you're planning on making some docs for how to use this.  Are people meant to store these on-chain?

File: libraries/concurrent-merkle-tree/src/state.rs
Comment (joncinque) [inline libraries/concurrent-merkle-tree/src/state.rs:6]: Same point here on the `Pod` / `Zeroable`, is it possible?

File: libraries/concurrent-merkle-tree/src/state.rs
Comment (joncinque) [inline libraries/concurrent-merkle-tree/src/state.rs:13]: Not sure if this makes any difference, but `Path` and `ChangeLog` are exactly the same struct. Is that a coincidence or is something much more beautiful at work?

File: libraries/concurrent-merkle-tree/src/cmt.rs
Comment (joncinque) [inline libraries/concurrent-merkle-tree/src/cmt.rs:60]: Considering all of the inner types should be safe to implement `Pod` / `Zeroable`, you should be good to go with this one too and avoid the `unsafe`!

**Review by joncinque — COMMENTED** [review]
Really great work! I'll be perfectly honest, I didn't dig into the index math on the interior structs. Most of my comments are small points around the overall structure of the project and types.  Maybe some other time I'll dig further in :smile:

File: libraries/merkle-tree-reference/src/lib.rs
Comment (jarry-xiao) [reply libraries/merkle-tree-reference/src/lib.rs:162]: `merkle-tree-reference` is only used for test cases, we can keep this internal!

File: libraries/concurrent-merkle-tree/src/cmt.rs
Comment (jarry-xiao) [reply libraries/concurrent-merkle-tree/src/cmt.rs:16]: Due to the bit math. We use `u32`

File: libraries/concurrent-merkle-tree/src/cmt.rs
Comment (jarry-xiao) [reply libraries/concurrent-merkle-tree/src/cmt.rs:74]: You can't `#[derive(Default)]` on a `[Foo; N]` where N is generic (might be impossible in general)

File: libraries/concurrent-merkle-tree/src/state.rs
Comment (jarry-xiao) [reply libraries/concurrent-merkle-tree/src/state.rs:6]: Impossible for const generics

File: libraries/concurrent-merkle-tree/src/state.rs
Comment (jarry-xiao) [reply libraries/concurrent-merkle-tree/src/state.rs:13]: They are 100% equivalent. The duplication is so the code is less confusing semantically. `path != proof`

File: libraries/concurrent-merkle-tree/src/state.rs
Comment (jarry-xiao) [reply libraries/concurrent-merkle-tree/src/state.rs:84]: I actually think there are no 128-bit architectures (to my knowledge) but there is 128-bit alignment (for u128s on aarch64). This is needed for the unsafe bytemuck impl to have the memory addresses of each `Path` fall on valid offsets (0 mod std::mem::align_of::<Path>() which should be 8 in this case)

File: libraries/concurrent-merkle-tree/src/cmt.rs
Comment (jarry-xiao) [reply libraries/concurrent-merkle-tree/src/cmt.rs:42]: Yes

File: libraries/concurrent-merkle-tree/src/cmt.rs
Comment (jarry-xiao) [reply libraries/concurrent-merkle-tree/src/cmt.rs:60]: Mentioned in the previous comment, but Pod / Zeroable unfortunately doesn't work for types with const generic parameters. It probably can if the bytemuck library is patched to do so. There's a lot of tricky edge cases to consider with architecture specific alignment though. For example:
```
struct Foo {
    a: u64,
    b: u128,
    c: u64,
}
```
is size 32 on most x86 and 48 on aarch64, because `b` needs to always fall on an address that is a multiple of 16.

This makes things like
```
struct Bar<T, const N: usize> {
    a: [T, N]
}
```
quite difficult to generalize, but it's certainly possible to have the library execute different codepaths based on the target_arch

File: libraries/concurrent-merkle-tree/src/state.rs
Comment (joncinque) [reply libraries/concurrent-merkle-tree/src/state.rs:84]: Ah gotcha, you mentioned 128-bit systems during our last chat about this. probably don't need to worry about it then

File: libraries/concurrent-merkle-tree/src/state.rs
Comment (joncinque) [reply libraries/concurrent-merkle-tree/src/state.rs:6]: Ah what a shame, it looks like a generic that gives a size is treated the same as any other generic

File: libraries/concurrent-merkle-tree/src/cmt.rs
Comment (joncinque) [reply libraries/concurrent-merkle-tree/src/cmt.rs:60]: Yeah, looks like you'd have to declare them as packed to make it work: https://github.com/Lokathor/bytemuck/issues/70 -- might even be safer to do that!

Comment (jarry-xiao) [reply libraries/concurrent-merkle-tree/src/cmt.rs:60]: Not sure how well this meshes with the rBPF virtual machine. Certainly not an expert, but historically I have not had a good time interacting with objects that are not byte aligned. I think this unsafe impl is completely safe on 64-bit systems like the Solana VM given that everything is padded correctly. There’re weird issues that arise when mutating a packed but unaligned object in a view that this completely side steps

Comment (joncinque) [reply libraries/concurrent-merkle-tree/src/cmt.rs:60]: Ah gotcha, no worries then

File: libraries/merkle-tree-reference/Cargo.toml
Comment (ngundotra) [reply libraries/merkle-tree-reference/Cargo.toml:3]: If there's any expressed interest, I am open to cleaning up the reference library to be used for any use case. For now, I think it's better to just keep it as a test crate.

File: libraries/merkle-tree-reference/Cargo.toml
Comment (ngundotra) [reply libraries/merkle-tree-reference/Cargo.toml]: Addressed

File: libraries/concurrent-merkle-tree/src/cmt.rs
Comment (ngundotra) [reply libraries/concurrent-merkle-tree/src/cmt.rs]: Addressed

File: libraries/concurrent-merkle-tree/src/error.rs
Comment (ngundotra) [reply libraries/concurrent-merkle-tree/src/error.rs]: Great catch, removed

File: libraries/concurrent-merkle-tree/Cargo.toml
Comment (ngundotra) [reply libraries/concurrent-merkle-tree/Cargo.toml]: Addressed

File: libraries/concurrent-merkle-tree/src/lib.rs
Comment (ngundotra) [reply libraries/concurrent-merkle-tree/src/lib.rs]: Links like this actually refer to the underlying file.

ie `[concurrent_merkle_tree]` links to `spl_concurrent_merkle_tree/concurrent_merkle_tree/index.html` in the docs

File: libraries/concurrent-merkle-tree/src/utils.rs
Comment (ngundotra) [reply libraries/concurrent-merkle-tree/src/utils.rs]: Split up files to address this comment :)

File: libraries/concurrent-merkle-tree/src/cmt.rs
Comment (ngundotra) [reply libraries/concurrent-merkle-tree/src/cmt.rs]: Added a comment

File: libraries/concurrent-merkle-tree/src/cmt.rs
Comment (ngundotra) [reply libraries/concurrent-merkle-tree/src/cmt.rs]: Yes this can fail. This is complex to reason about in the practical case, so I expect this function to change quite a bit in the coming months.

This now returns `InvalidProof`

File: libraries/concurrent-merkle-tree/src/cmt.rs
Comment (ngundotra) [reply libraries/concurrent-merkle-tree/src/cmt.rs]: Great catch. Updated to return `()`.

All methods in the public interface initially returned a `Node` type to make the contract-level macro easier to write. We've since fixed this code-smell, so this function can return `()` as expected.

File: libraries/concurrent-merkle-tree/src/cmt.rs
Comment (ngundotra) [reply libraries/concurrent-merkle-tree/src/cmt.rs]: Great catch, updated

File: libraries/concurrent-merkle-tree/src/cmt.rs
Comment (ngundotra) [reply libraries/concurrent-merkle-tree/src/cmt.rs]: Addressed

File: libraries/concurrent-merkle-tree/src/cmt.rs
Comment (ngundotra) [reply libraries/concurrent-merkle-tree/src/cmt.rs]: Addressed

File: libraries/concurrent-merkle-tree/src/error.rs
Comment (ngundotra) [reply libraries/concurrent-merkle-tree/src/error.rs]: Addressed

**Comment by joncinque** [comment]
Awesome, thanks for addressing all the points!

**Review by joncinque — APPROVED** [review]
I'll let you do the honors and merge it

**Pull request merged**
